### PR TITLE
Make admin change two-step

### PR DESCRIPTION
**Pull request by CodeSandwich**
This should make it more difficult to accidentally lose admin. This role is so critical that it probably makes sense to do that, all other serious errors and mistakes can be fixed by the admin.

### CODE REVIEW

File: src/Managed.sol [hunk 1]
```diff
@@ -81,10 +89,36 @@ abstract contract Managed is UUPSUpgradeable {
...
+    /// Sets the proposed admin to the zero address.
+    /// Can only be called by the proposed admin.
+    function acceptAdmin() public {
+        require(proposedAdmin() == msg.sender, "Caller not the proposed admin");
+        _managedStorage().proposedAdmin = address(0);
+        _changeAdmin(msg.sender);
+    }
+
+    /// @notice Changes the admin of the contract to address zero.
+    /// It's no longer possible to upgrade the contract afterwards.
```
Comment (xmxanuel) [inline src/Managed.sol:117]:
```suggestion
    /// It's no longer possible to upgrade the contract afterwards if no proposed admin has been set
```
I could still propose another admin before I renounce

File: src/Managed.sol [hunk 2]
```diff
@@ -81,10 +89,36 @@ abstract contract Managed is UUPSUpgradeable {
...
+    function proposedAdmin() public view returns (address) {
+        return _managedStorage().proposedAdmin;
+    }
+
+    /// @notice Proposes a change of the admin of the contract.
+    /// The proposed new admin must call `acceptAdmin` to finalize the change.
+    /// To cancel a proposal propose a different address, e.g. the zero address.
+    /// Can only be called by the current admin.
+    /// @param newAdmin The proposed admin address.
+    function proposeAdmin(address newAdmin) public onlyAdmin {
```
Comment (xmxanuel) [inline src/Managed.sol:102]:
```suggestion
    function proposeNewAdmin(address newAdmin) public onlyAdmin {
```
Maybe?
`proposeAdmin` and `proposedAdmin` are so simliar.

File: src/Managed.sol [hunk 3]
```diff
@@ -81,10 +89,36 @@ abstract contract Managed is UUPSUpgradeable {
...
+    }
+
+    /// @notice Changes the admin of the contract to address zero.
+    /// It's no longer possible to upgrade the contract afterwards.
     /// Can only be called by the current admin.
-    function changeAdmin(address newAdmin) public onlyAdmin {
-        _changeAdmin(newAdmin);
+    function renounceAdmin() public onlyAdmin {
+        emit AdminChanged(admin(), address(0));
+        StorageSlot.getAddressSlot(_ADMIN_SLOT).value = address(0);
```
Comment (xmxanuel) [inline src/Managed.sol:121]: I just looked it up why you are not using the existing function.

```solidity
    function _setAdmin(address newAdmin) private {
        require(newAdmin != address(0), "ERC1967: new admin is the zero address");
        StorageSlot.getAddressSlot(_ADMIN_SLOT).value = newAdmin;
    }

    /**
     * @dev Changes the admin of the proxy.
     *
     * Emits an {AdminChanged} event.
     */
    function _changeAdmin(address newAdmin) internal {
        emit AdminChanged(_getAdmin(), newAdmin);
        _setAdmin(newAdmin);
    }
```

:)

File: src/Managed.sol (hunk 2 above)
Comment (CodeSandwich) [reply src/Managed.sol]: Good idea, applied.

**Comment by CodeSandwich** [comment]
I think that we don't need `denyProposedAdmin`. It's basically the same as `proposeNewAdmin(address(0))`, the added explicitness is nice, but it's probably not worth the ABI surface, plus we're really stretching the contract bytecode size and each public function takes space.

**Pull request merged**